File: memento/src/main/java/com/alexstyl/specialdates/events/namedays/calendar/resource/GreekSpecialNamedays.java
package com.alexstyl.specialdates.events.namedays.calendar.resource;

import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.events.namedays.NameCelebrations;
import com.alexstyl.specialdates.events.namedays.NamesInADate;
import com.alexstyl.specialdates.events.namedays.calendar.OrthodoxEasterCalculator;

import java.util.ArrayList;

final class GreekSpecialNamedays implements SpecialNamedays {

    private final GreekNamedays greekNamedays;

    public static GreekSpecialNamedays from(NamedayJSON namedayJSON, OrthodoxEasterCalculator easterCalculator) {
        GreekNamedays greekNamedays = GreekNamedays.from(namedayJSON.getSpecial(), easterCalculator);
        return new GreekSpecialNamedays(greekNamedays);
    }

    private GreekSpecialNamedays(GreekNamedays greekNamedays) {
        this.greekNamedays = greekNamedays;
    }

    @Override
    public NamesInADate getNamedayOn(Date date) {
        return greekNamedays.getNamedayByDate(date);
    }

    @Override
    public ArrayList<String> getAllNames() {
        return greekNamedays.getNames();
    }

    @Override
    public NameCelebrations getNamedaysFor(String name, int year) {
        return greekNamedays.getNamedaysFor(name, year);
    }
}

File: memento/src/main/java/com/alexstyl/specialdates/events/namedays/calendar/resource/GreekSpecialNamedays.kt
package com.alexstyl.specialdates.events.namedays.calendar.resource

import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.events.namedays.NameCelebrations
import com.alexstyl.specialdates.events.namedays.NamesInADate
import com.alexstyl.specialdates.events.namedays.calendar.OrthodoxEasterCalculator

internal class GreekSpecialNamedays private constructor(private val greekNamedays: GreekNamedays) : SpecialNamedays {

    override val allNames: List<String>
        get() = greekNamedays.names

    override fun getNamedayOn(date: Date): NamesInADate {
        return greekNamedays.getNamedayByDate(date)
    }

    override fun getNamedaysFor(name: String, year: Int): NameCelebrations {
        return greekNamedays.getNamedaysFor(name, year)
    }

    companion object {

        fun from(namedayJSON: NamedayJSON, easterCalculator: OrthodoxEasterCalculator): GreekSpecialNamedays {
            val greekNamedays = GreekNamedays.from(namedayJSON.special, easterCalculator)
            return GreekSpecialNamedays(greekNamedays)
        }
    }
}
